ui/pages/api/etherscan/wallet-tokens.ts

```diff
@@ -0,0 +1,194 @@
+  }
+> = new Map()
+
+const CACHE_DURATION = 5 * 60 * 1000 // 5 minutes cache
+
+// Function to detect if a token name/symbol contains URL patterns (likely scam)
+function containsUrlPattern(text: string): boolean {
+  if (!text) return false
+
+  const urlPatterns = [
+    // Domain extensions
+    /\.(com|io|org|net|co|me|ly|to|cc|tk|ml|ga|cf|info|biz|xyz|top|click|online|site|website|app|tech|crypto|finance|exchange|swap|defi|nft)(\b|$)/i,
+    // Protocol patterns
+    /(https?:\/\/|www\.)/i,
+    // Short link services
+    /(bit\.ly|t\.ly|tinyurl|short\.link|rebrand\.ly|cutt\.ly|is\.gd|ow\.ly)/i,
+    // Common scam patterns
+    /(\w+\.\w{2,4}\/\w+)/i, // domain/path pattern
+  ]
+
+  return urlPatterns.some((pattern) => pattern.test(text))
+}
+
+// Supported chain IDs for Etherscan V2 API
+const SUPPORTED_CHAINS: Record<string, number> = {
+  ethereum: 1,
+  polygon: 137,
+  bsc: 56,
+  arbitrum: 42161,
+  optimism: 10,
+  base: 8453,
+  avalanche: 43114,
+  sepolia: 11155111,
+  'arbitrum-sepolia': 421614,
+  'optimism-sepolia': 11155420,
+}
+
+async function fetchTokenBalance(
+  contractAddress: string,
+  walletAddress: string,
+  chainId: number,
+  apiKey: string
+): Promise<string> {
+  try {
+    const balanceUrl = `https://api.etherscan.io/v2/api?chainid=${chainId}&module=account&action=tokenbalance&contractaddress=${contractAddress}&address=${walletAddress}&tag=latest&apikey=${apiKey}`
+
+    const response = await fetch(balanceUrl)
+    const data = await response.json()
+
+    if (data.status === '1') {
+      return data.result || '0'
+    }
+    return '0'
+  } catch (error) {
+    return '0'
+  }
+}
+
+async function handler(req: NextApiRequest, res: NextApiResponse) {
+  try {
+    const { address, chain = 'ethereum', page = '1', offset = '50' } = req.query
+
+    // Validate required parameters
+    if (!address || typeof address !== 'string') {
+      return res.status(400).json({ error: 'Wallet address is required' })
+    }
+
+    if (!SUPPORTED_CHAINS[chain as string]) {
+      return res.status(400).json({
+        error: `Unsupported chain. Supported chains: ${Object.keys(
+          SUPPORTED_CHAINS
+        ).join(', ')}`,
+      })
+    }
+
+    const chainId = SUPPORTED_CHAINS[chain as string]
+    const cacheKey = `${address}-${chain}-${page}-${offset}`
+    const now = Date.now()
+
+    // Check cache
+    const cached = tokenCache.get(cacheKey)
+    if (cached && now - cached.timestamp < CACHE_DURATION) {
+      return res.status(200).json(cached.data)
+    }
+
+    const apiKey = process.env.NEXT_PUBLIC_ETHERSCAN_API_KEY
+
+    if (!apiKey) {
+      return res.status(500).json({
+        error: 'Etherscan API key not configured',
+      })
+    }
+
+    // Step 1: Get token transfer events to discover tokens using V2 API
+    const tokentxUrl = `https://api.etherscan.io/v2/api?chainid=${chainId}&module=account&action=tokentx&address=${address}&page=1&offset=10000&sort=desc&apikey=${apiKey}`
```

Comment:
can't believe get tokens held by an address is a PRO endpoint. smh.

definitely did not used to be like that